Add a new prop to card in order to flex a class for a new style

alex-bezek opened the pull request:
### Summary
New prop called variant used to flex to different card styles for new use cases. Falls back to default. Currently only adds the option for raised which adds a new css custom attribute for the box shadow.

### Additional Details
Resolves #1432
Deployed https://terra-core-deployed-pr-1458.herokuapp.com/#/site/components/site/card/index

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

alex-bezek commented on packages/terra-card/src/Card.scss:
```diff
@@ -3,8 +3,15 @@
+  }
+
+  .raised {
+    box-shadow: var(--terra-card-raised-box-shadow, none);
```
@neilpfeiffer let me know if you have a style you want applied by default for this class in terra-core. Otherwise, it's just a new themeable option we can utilize

neilpfeiffer replied:
@alex-bezek - Yes, we will want to have a visual version for default theme, otherwise it would look like a dead prop (not recommended). I'll grab a box-shadow style to use.

neilpfeiffer replied:
@alex-bezek Values to used for default terra-opensource-theme:
```
background: white;
border: 1px solid #c8cacb; 
box-shadow: 0 3px 8px -2px rgba(0, 0, 0, 0.18), 0 3px 10px 0 rgba(0, 0, 0, 0.12), 0 3px 3px -3px rgba(0, 0, 0, 0.3);
```

alex-bezek commented on packages/terra-card/src/Card.jsx:
```diff
@@ -12,20 +12,31 @@ const propTypes = {
+  /**
+   * An option used in order to change to a different style of card for different use cases
+   */
+  variant: PropTypes.oneOf(['default', 'raised']),
```
I couldn't find a pattern on what the prefered way to do this is. I see
* just an array inline like this
* a const array up above and then used here
* an object where the keys and values are all caps and then doing object.values here
* adding `opts` to the component, exporting, and using those.

Let me know if there is a preference!

dylan-sftwr replied:
I'm currently working on #1044. I volunteered to go through all the components and do the uplift to allow for exported enumeration of variants, I'd prefer it be done following [this](https://github.com/cerner/terra-core/commit/8bae7969950a3868334cf871b72da372dea373ee) pattern, as it's how I plan on implementing all component enumerations.

dev-cprice replied:
I like the use of enums, but I think there needs to be a conversation on the consumption of them. Because it's easier and more terse to do something like `<Card variant="raised" ... />` than `<Card variant={Card.Opts.Variants.RAISED} ... />` as a consumer. I get why the precedence is to put it behind another `Opts` object, but it just adds to verbosity and feels like writing Java all over again. 

`Card.Variants.RAISED` >>> `Card.Opts.Variants.RAISED`, in my opinion

You could also name export `Variants` or `CardVariants` from here, but `CardVariants` is only one key stroke less than `Card.Variants` and `Variants` will need to be renamed by the consumer if they use multiple terra components in the same file (all too common), e.g. `import { Variants as CardVariants } from 'terra-card';

The above can also be done with destructuring from the Card like `const { Variants } = Card` or renaming it like `const { Variants: CardVariants } = Card;`

I don't putting the burden of renaming an enum on the consumer and I think if getting to the enum takes too long a consumer is just going to do `<Card variant="raised" ... />`

alex-bezek replied:
Updated

emilyrohrbough commented on packages/terra-card/src/Card.jsx:
```diff
@@ -12,20 +12,31 @@ const propTypes = {
    */
   children: PropTypes.node,
+  /**
+   * An option used in order to change to a different style of card for different use cases
```
Let list the options in the description so consumers know what variants there are.

alex-bezek replied:
Good call, these are what generate the doc site props table, aren't they?

emilyrohrbough commented on packages/terra-card/tests/jest/Card.test.jsx:
```diff
@@ -14,6 +14,13 @@ describe('Card', () => {
+    expect(wrapper.find('.default'));
+  });
+
+  it('should use the raised variant when prop is provided', () => {
```
Lets add a jest test using the `default` variant as well

JakeLaCombe commented on packages/terra-card/tests/jest/Card.test.jsx:
```diff
@@ -14,6 +14,13 @@ describe('Card', () => {
+
+  it('should use the raised variant when prop is provided', () => {
+    const wrapper = shallow(<Card variant="raised" />);
+    expect(wrapper.find('.raised'));
```
Do we need the wrapper.find tests? Your snapshot had a card with the raised class on it. To me, it feels like duplicate testing when we do test jest like this.

alex-bezek replied:
My personal opinion is that snapshot tests are not very helpful and often times ignored. I typically prefer explicit jest tests. 
That being said, i can remove the test if you want

mjhenkes commented on packages/terra-card/src/Card.jsx:
```diff
@@ -6,26 +6,47 @@ import styles from './Card.scss';
 Card.propTypes = propTypes;
+Card.defaultProps = defaultProps;
 Card.Body = CardBody;
+Card.Opts = opts;
```
For the record, I'm against exporting these props as an enum. Using a string is more concise and lowers the complexity of maintaining and creating these components. With prop type warning you will be warned if you've had a spelling mistake or pass in a non supported type.

alex-bezek replied:
For the record, I agree with you. The first is easier
`<Card variant="raised" />`
vs
`<Card variant={Card.Opts.Variants.RAISED} />`

I also ran my wdio tests for 15 min only for them to fail because I used
`<Card variant={Card.Opts.RAISED} />` the first time on accident

Also it makes typing the wrong thing way worse. With the string it will just throw a prop type warning in dev and continue along its way. But with the object of object of object thing, mis typing like
`<Card id="raisedCard" variant={Card.Opt.Variants.RAISED}>`
throws a run time exception and crashes the page

![image](https://user-images.githubusercontent.com/8029578/39439716-2315f0c8-4c6e-11e8-9b37-658acb75e8eb.png)

or `<Card id="raisedCard" variant={Card.Opts.Variant.RAISED}>`
![image](https://user-images.githubusercontent.com/8029578/39439759-53775752-4c6e-11e8-9cf0-e1af99d4d171.png)

But, I don't really have a dog in this. Let me know what the consensus is to go with, and it sounds like there is an open issue to reconcile them all to be consistent anyways.

dylan-sftwr commented:
@mjhenkes if @alex-bezek just wants to remove the enumeration from this PR and do it with strings, I can just pick it up in #1044 if we decide to go forward with it.

mjhenkes commented:
@dylan-sftwr Sounds good. Lets not block this PR.

alex-bezek commented:
Updated to use an object inside the component, but not export it out. 
Also updated the default from none to the box shadow the @neilpfeiffer provided.
Deployment link here: https://terra-core-deployed-pr-1458.herokuapp.com/#/site/components/site/card/index

Let me know if there are any other issues

neilpfeiffer commented:
Tested: PR #1458
Resolves: Issue #1432

**Reviewed:**
- Reviewed Enhancement to add an additional card variant style: `raised`
- Follows pattern considerations documented in Cerner Design Standards: [Cards](https://wiki.ucern.com/x/JG4VYQ)
- Follows UX defined design implementation for Terra-Card 
- Follows all defined theme designs: terra-theme

\
Verified by @neilpfeiffer and @scottwilmarth, Ready for Release after Functional Validation.


**Pass** |
-----------|
UX Review|
